selfdrive/boardd/boardd.cc

```diff
@@ -165,27 +176,116 @@ bool usb_connect() {
   }
 
   connected_once = true;
-  panda = tmp_panda.release();
+  if (tmp_panda->hw_type != cereal::PandaState::PandaType::UNO && tmp_panda->hw_type != cereal::PandaState::PandaType::DOS) {
+    panda_aux = tmp_panda.release();
+  } else {
+    panda = tmp_panda.release();
+  }
   return true;
 }
 
 // must be called before threads or with mutex
 static bool usb_retry_connect() {
-  LOGW("attempting to connect");
-  while (!do_exit && !usb_connect()) { util::sleep_for(100); }
+  
+  LOGW("attempting to find boards");
+  size_t detect_count = 0;
+  while (detect_count == 0) {
+    util::sleep_for(100);
+    pandas_detected = Panda::list();
+    detect_count = pandas_detected.size();
+  }
+  
+  for(int i=0; i<detect_count; ++i) {
+    LOGW("board #%d USB serial: %s", i, pandas_detected[i].c_str());
+  }
+  LOGW("total boards detected: %d", detect_count);
+
+  while (!do_exit && (detect_count != pandas_connected.size()) ) {
+    usb_connect();
+    util::sleep_for(100);
+  }
   if (panda) {
-    LOGW("connected to board");
+    LOGW("connected to main board: %s", panda->usb_serial.c_str());
+  } else { return false; }
+  if (panda_aux) {
+    LOGW("connected to aux board: %s", panda_aux->usb_serial.c_str());
   }
   return !do_exit;
 }
 
 void can_recv(PubMaster &pm) {
-  kj::Array<capnp::word> can_data;
-  panda->can_receive(can_data);
+
+  uint32_t data[(RECV_SIZE/4)*2]; // Double the size, for both pandas
+  size_t num_msg;
+
+  num_msg = panda->can_receive_raw(&data[num_msg*4], main_shift);
+  if (panda_aux != nullptr) num_msg += panda_aux->can_receive_raw(&data[num_msg*4], aux_shift);
+
+  MessageBuilder msg;
+  auto evt = msg.initEvent();
+  evt.setValid(panda->comms_healthy); // Should be comms_healthy from each panda?
+
+  // populate message
+  auto canData = evt.initCan(num_msg);
+  for (int i = 0; i < num_msg; i++) {
+    canData[i].setAddress(data[i*4]);
+    canData[i].setBusTime(data[i*4+1] >> 16);
+    int len = data[i*4+1]&0xF;
+    canData[i].setDat(kj::arrayPtr((uint8_t*)&data[i*4+2], len));
+    canData[i].setSrc((data[i*4+1] >> 4) & 0xff);
+  }
+  auto can_data = capnp::messageToFlatArray(msg);
   auto bytes = can_data.asBytes();
   pm.send("can", bytes.begin(), bytes.size());
 }
 
+void split_messages(capnp::List<cereal::CanData>::Reader can_data_list) {
```

Comment:
Why do we need to split the messages? Maybe it's better to move this into the panda class as well. When initializing a panda object you can give it a list of busses to use, and an optional shift. Then you can pass the full list of messages into the `can_send` function of both pandas.

Reply:
I like that idea, thank you!